### DISCUSSION TITLE
How to start automatically caddy as soon as Android starts

### DISCUSSION THREAD

**Discussion by Marietto2008** [termux/termux-app#4105] [General]
Hello.

Is there a method to start automatically a command based on caddy as soon as Android (14) starts ? At the moment I need to start termux and then I should write the command :

`sudo caddy run -c /data/Caddyfile`

and this is uncomfortable. Thanks.

**Comment by agnostic-apollo** [comment]
https://github.com/termux/termux-boot

**Comment by Marietto2008** [comment]
it does not work. Can you help me again ? thanks. This is the script that I need to run at boot :

```
termux-wake-lock
sudo caddy run -c /data/data/com.termux/files/home/.termux/boot/Caddyfile
```

as soon as Android boot,that script is not executed. Probably because I should run it with sudo. 
Unfortunately if I don't run it with sudo,it is not able to work properly as reverse_proxy,probably for a permission problem.
Do you know how can I run it at boot with sudo ? thanks.

**Reply by agnostic-apollo** [reply]
1. Did you install the termux-boot app?
2. Did you unlock the lock screen and wait like a minute or more?
3. Did you create the script directory or file with root?

**Reply by Marietto2008** [reply]
1. yes,I've installed the apk that I found here : https://f-droid.org/repo/com.termux.boot_1000.apk
2. I have removed any lock screen from the Android settings
3. yes,I can create files and folders only if I'm root. So I rooted Android.

Probably my problem is that I need to find a way to run the commands inside the script using sudo. Can you check if the script is ran when it contains a command prefixed by sudo ?

**Reply by agnostic-apollo** [reply]
Never create directories and files with root that are used by Termux app or normal processes in home or prefix. Restore attributes 

https://github.com/termux/termux-app/issues/2080#issuecomment-846575028

Also you can look for errors in logcat.

https://github.com/termux/termux-packages/blob/site/site/pages/en/projects/docs/repos/main/packages/termux-tools/logcat/index.md

**Reply by Marietto2008** [reply]
Ok. I've restored the permissions as they should be,but the script start-caddy is not executed as soon as Android boot.

This is the scenario :

1) the script tries to run caddy without sudo :

```
/data/data/com.termux/files/home/.termux/boot/start-caddy :

/data/data/com.termux/files/usr/bin/termux-wake-lock
/data/data/com.termux/files/usr/bin/caddy run -c /data/data/com.termux/files/home/.termux/boot/Caddyfile 

./start-caddy 

INFO : using config from file "file" "/data/data/com.termux/files/home/.termux/boot/Caddyfile
INFO : adapted config to JSON "adapter": "caddyfile"
WARN : Caddyfile input is not formatted; run "caddy fmt --overwrite" to fix inconsistencies "adapter" ; "caddyfile" ; "file": "/data/data/com.termux/files/home/.termux/boot/Caddyfile", "line" : 2
INFO : admin admin endpoint started "address" ; "localhost:2019" , "enforce_origin" : false, "origins" : [//localhost:2019", "//[::1] 2019" , "//127.0.0.1:2019"

INFO : tls cache maintenance started background certificate maintenance "cache" : "0x400"
INFO : http auto_https server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS "server_name" : "srv0" , "https_port" : 443
INFO : http auto_https enabling automatic HTTP-HTTPS redirects "server_name": "srv0"
INFO : tls.cache.maintenance stopped background certificate maintenance "cache" : "0x400"
ERROR : loading initial config: loading new config: http app mobile start : listening on :443 listen tcp: 443 bind: permission denied
```


2)  the script tries to run caddy with sudo :


```
/data/data/com.termux/files/home/.termux/boot/start-caddy :

/data/data/com.termux/files/usr/bin/termux-wake-lock
/data/data/com.termux/files/usr/bin/sudo caddy run -c /data/data/com.termux/files/home/.termux/boot/Caddyfile

./start-caddy

INFO : using config from file "file" "/data/data/com.termux/files/home/.termux/boot/Caddyfile
INFO : adapted config to JSON "adapter": "caddyfile"
WARN : Caddyfile input is not formatted; run "caddy fmt --overwrite" to fix inconsistencies "adapter" ; "caddyfile" ; "file": "/data/data/com.termux/files/home/.termux/boot/Caddyfile", "line" : 2
INFO : admin admin endpoint started "address" ; "localhost:2019" , "enforce_origin" : false, "origins" : [//localhost:2019", "//[::1] 2019" , "//127.0.0.1:2019"
INFO : http auto_https server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS "server_name" : "srv0" , "https_port" : 443
INFO : http auto_https enabling automatic HTTP-HTTPS redirects "server_name": "srv0"
INFO : tls.cache.maintenance started background certificate maintenance "cache" : "0x400"
INFO : http enabling HTTP/3 listener addr 443
INFO : http log server running "name" ; "srv0" ; "protocols" ; [h1 ; h2 ; h3]
INFO : http log server running "name" ; "remaining_auto_https_redirects" ; "protocols" [h1,h2,h3]
INFO : http enabling automatic TLS certificate management "domains" "ziomario.ns0.it"
INFO : autosaved config (load with --resume flag) "file": "/data/data/com.termux/files/home/.suroot/.config/caddy/autosave.json"
INFO : serving initial configuration
INFO : tls storage cleaning happened too recently; skipping for now "storage": "FileStorage: /data/data/com.termux/files/home/.suroot/caddy" , "instance": "312 bla bla" , "try_again" : "2024/08/17 10:04" , "try_again_in" : 86399
INFO : tls finished cleaning storage units
```

in both cases,it is not executed when Android starts,but the version that contains sudo is executed after Android started,by opening termux and writing : ./start-caddy.

**Reply by agnostic-apollo** [reply]
Wait at least 1-2mins after boot (or lockscreen unlock) and see if script executes. And logcat should be able to tell you what is happening. Search for termux entries after getting **system log**.

https://github.com/termux/termux-packages/blob/site/site/pages/en/projects/docs/repos/main/packages/termux-tools/logcat/index.md#to-get-full-system-logs-of-other-apps-and-the-android-system

**Reply by Marietto2008** [reply]
I've sent my logcat log file to your email,if you want to give a look. Im not able to understand what's broken and I can't paste here a lot of log entries.

**Reply by agnostic-apollo** [reply]
>W linker64: type=1400 audit(0.0:19964): avc:  granted  { execute } for  path="/data/data/com.termux/files/usr/bin/runsv" dev="mmcblk0p4" ino=9568721 scontext=u:r:untrusted_app:s0:c125,c256,c512,c768 tcontext=u:object_r:app_data_file:s0:c125,c256,c512,c768 tclass=file app=com.termux

One main reason that boot tasks are failing is because you have installed the Termux app from PlayStore and Termux:Boot app from F-Droid, they are not compatible. Your Termux:Boot app will not have access to `/data/data/com.termux` directory as Termux app on PlayStore is not using `sharedUserId`, like Termux:Boot is. Install all Termux apps from F-Droid/GitHub. If you want to use the one from PlayStore, then report any issues to https://github.com/termux-play-store/termux-apps as we won't provide support for it, read #4000 and https://github.com/termux/termux-app#installation

https://github.com/termux/termux-boot/blob/master/app/src/main/java/com/termux/boot/BootReceiver.java#L25

**Reply by Marietto2008** [reply]
Bingo. Finally it worked. I wonder why the developer of the Termux app from the Playstore didn't fix the bug in his application.

**Comment by Marietto2008** [comment]
I've saved the script below inside the directory /data/data/com.termux/files/home/.termux/boot :

```
start-caddy :

/data/data/com.termux/files/usr/bin/termux-wake-lock
/data/data/com.termux/files/usr/bin/sudo caddy run -c /data/data/com.termux/files/home/.termux/boot/Caddyfile
```

and I did : 

`chmod +x start-caddy`

the script is not executed at all as soon as Android starts. But it works great if I run it in this way :

`./start-caddy`

where is the problem now ?
